src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/FairySouls.java

```diff
@@ -21,44 +21,273 @@
 import net.minecraftforge.event.world.WorldEvent;
 import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
 
-import java.io.*;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
 import java.lang.reflect.Type;
 import java.nio.charset.StandardCharsets;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.TreeMap;
 import java.util.stream.Collectors;
 
 public class FairySouls {
+	private static FairySouls instance = null;
 	private static final String unknownProfile = "unknown";
-	private static List<BlockPos> currentSoulList = null;
-	private static List<BlockPos> currentSoulListClose = null;
-	private static HashMap<String, HashMap<String, Set<Integer>>> loadedFoundSouls = new HashMap<>();
 
-	private static HashMap<String, Set<Integer>> getFoundSoulsForProfile() {
+	private boolean trackSouls;
+	private boolean showSouls;
+	private HashMap<String, HashMap<String, Set<Integer>>> allProfilesFoundSouls = new HashMap<>();
+	private List<BlockPos> allSoulsInCurrentLocation;
+	private Set<Integer> foundSoulsInLocation;
+	private TreeMap<Double, BlockPos> missingSoulsDistanceSqMap;
```

Comment:
I don't really like the idea of:

 - using `double` as a `Map` key.
 - storing the distance of missing souls in a map with the distance as a key (possible loss of souls because of duplicate keys is very unlikely, but this still *feels* wrong)
 - using the implicit sorting of `TreeMap` to sort a collection.

None of those are a deal breaker, but since we are using this only in one method (minus initialization), i would at least inline this variable to contain whatever this is.

Reply:
This PR tries to strike the right balance between cleaning up the code, adding the desired features, and minimizing the chance of regression. While I agree with your comment, I chose not to change this part of the design as part of this PR. I just renamed / moved the existing logic; see line 148 in the original file where distanceSqMap was defined.